kernel/src/process/kill.rs

```diff
@@ -131,113 +131,57 @@ pub fn kill_all(signal: Option<UserSignal>, ctx: &Context) -> Result<()> {
-
-        // FIXME: If the thread is exiting concurrently, it may miss the signal queued on it.
-        if thread.is_exited() {
-            continue;
-        }
-
-        if !posix_thread.has_signal_blocked(*signum) {
-            // Send the signal to any alive thread that does not block the signal.
-            thread_to_enqueue = Some(posix_thread);
-            break;
-        } else if thread_to_enqueue.is_none() {
-            // If all threads block the signal, send it to the first permitted and alive thread.
-            // FIXME: If it exits later with the signals still blocked, it will miss the signal
-            // queued on it.
-            thread_to_enqueue = Some(posix_thread);
-        }
-    }
+    let target_main_thread = process.main_thread();
+    check_signal_perm(target_main_thread.as_posix_thread().unwrap(), ctx, signum)?;
 
-    if !found_permitted_thread {
-        return_errno_with_message!(
-            Errno::EPERM,
-            "the signal cannot be sent to the target process"
-        );
-    }
-
-    let Some(thread_to_enqueue) = thread_to_enqueue else {
-        // All threads have exited. This is a zombie process.
-        return Ok(());
+    if let Some(signal) = signal {
+        process.enqueue_signal(signal);
     };
 
-    // Since `thread_to_enqueue` has been set, `signal` cannot be `None`.
-    let signal = signal.unwrap();
-
-    // Drop the signal if it's ignored. See explanation at `enqueue_signal_locked`.
-    let signum = signal.num();
-    if sig_dispositions.get(signum).will_ignore(signum) {
-        return Ok(());
-    }
-
-    thread_to_enqueue.enqueue_signal_locked(Box::new(signal), &sig_dispositions);
-
     Ok(())
 }
 
-/// The IDs of the signal sender process.
-///
-/// For all signals, this structure includes the sender thread's effective user ID and real user
-/// ID. For SIGCONT, this structure additionally includes the sender thread's session ID.
-pub(super) struct SignalSenderIds {
-    ruid: Uid,
-    euid: Uid,
-    sid: Option<Sid>,
-}
+// Reference: <https://elixir.bootlin.com/linux/v6.17/source/kernel/signal.c#L799>.
+fn check_signal_perm(target: &PosixThread, ctx: &Context, signum: Option<SigNum>) -> Result<()> {
+    let target_process = target.process();
 
-impl SignalSenderIds {
-    pub(self) fn for_current_thread(ctx: &Context, signum: Option<SigNum>) -> Self {
-        let credentials = ctx.posix_thread.credentials();
-        let ruid = credentials.ruid();
-        let euid = credentials.euid();
-
-        let sid = signum.and_then(|signum| {
-            if signum == SIGCONT {
-                Some(ctx.process.sid())
-            } else {
-                None
-            }
-        });
-
-        Self { ruid, euid, sid }
+    if Arc::ptr_eq(&target_process, &ctx.process) {
+        return Ok(());
     }
 
-    pub(super) fn ruid(&self) -> Uid {
-        self.ruid
+    let current_cred = ctx.posix_thread.credentials();
+    let target_cred = target.credentials();
+
```

Comment:
Remove this blank line as the two `*_cred` variables are only used in the following `if` statement?